[Nov 15]  push only, force cleanstack and don't accept < 100 byte txns

ptschip opened the pull request.

deadalnix committed:
Make push only mandatory when magnetic anomaly activates.

Summary:
As per title.

Depends on D1622 and D1611

Test Plan: Added test cases in abc-magnetic-anomaly-activation

Reviewers: #bitcoin_abc, jasonbcox

Reviewed By: #bitcoin_abc, jasonbcox

Subscribers: jasonbcox, teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1623

Commit:
[qa] fix flaky abc-magnetic-anomaly-activation.py

Summary:
Currently, this test has a tranasction generator for a specific size.
It slowly grows the transaction padding to ensure it is exactly 101 bytes.
However, if the penultimate iteration of the loop generates a padding with
a leftmost bit of zero, and the ultimate run generates a padding with a
leftmost bit of 1, the test will fail.

This commit ensure the leftmost bit is always 1.

Test Plan:
Run in a loop for 100 runs:

  ./test/functional/test_runner.py abc-magnetic-anomaly-activation.py

Reviewers: #bitcoin_abc, deadalnix, jasonbcox

Reviewed By: #bitcoin_abc, deadalnix, jasonbcox

Subscribers: teamcity

Differential Revision: https://reviews.bitcoinabc.org/D1700

ptschip committed:
Impose Mimimum transaction size of 100 bytes after Nov15,2018 HF

deadalnix committed:
Add cleanstack check when activating November, 15 2018 upgrade

Summary:
As per suggestion by @dgenr8

Co-authored-by: Amaury Séchet <deadalnix@gmail.com>
Co-authored-by: Jason B. Cox <contact@jasonbcox.com>

Test Plan: Added test cases in qa/rpc-tests/nov152018_forkactivation.py

Reviewers: #bitcoin_abc, Mengerian, deadalnix

Reviewed By: #bitcoin_abc, Mengerian, deadalnix

Subscribers: Mengerian, teamcity, dgenr8

Differential Revision: https://reviews.bitcoinabc.org/D1647

ptschip committed:
Fork activation test update

sickpig commented on qa/rpc-tests/nov152018_forkactivation.py:
```diff
@@ -0,0 +1,274 @@
+                    tx.vout[2] = CTxOut(0, CScript(
+                        [random.getrandbits(8 * push_size - 1), OP_RETURN]))
+                    push_size += 1
+
```
see https://reviews.bitcoinabc.org/D1700 for a fix for possible spurious failure

ptschip replied:
thanks..done

sickpig commented on qa/rpc-tests/test_framework/blocktools.py:
```diff
@@ -59,6 +59,12 @@ def create_coinbase(height, pubkey = None):
         coinbaseoutput.scriptPubKey = CScript([OP_TRUE])
     coinbase.vout = [ coinbaseoutput ]
+
+    # Make sure the coinbase is at least 100 bytes
```
doens't make much of a difference but why not making the below change conditional to activation time param?

sickpig commented on src/miner.cpp:
```diff
@@ -163,8 +163,16 @@ CTransactionRef BlockAssembler::coinbaseTx(const CScript &scriptPubKeyIn, int _n
+    uint64_t nCoinbaseSize = ::GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION);
+    if (nCoinbaseSize < MIN_TX_SIZE)
+    {
+        tx.vin[0].scriptSig << std::vector<uint8_t>(MIN_TX_SIZE - nCoinbaseSize - 1);
```
ditto

awemany commented on qa/rpc-tests/nov152018_forkactivation.py:
```diff
@@ -0,0 +1,277 @@
+            return block
+
+        # shorthand for functions
+        block = self.next_block
```
I find this pretty confusing. `block` is used both as a `CBlock` as well as the method `next_block()`. Wouldn't it be better to make this explicit?

awemany commented on qa/rpc-tests/nov152018_forkactivation.py:
```diff
@@ -0,0 +1,277 @@
+        #check that this block was rejected.
+        yield rejected()
+        #yield rejected(RejectResult(16, b'bad-txns-undersize'))
+        assert_equal(bestblockhash, self.nodes[0].getbestblockhash())
```
Shouldn't this test the boundary - as in also make sure that `MIN_TX_SIZE` just barely goes through?

ptschip committed:
Padd the coinbase to be at least 100 bytes only after fork activation.

Previously it was enabled already by default.

ptschip committed:
Remove shadowing of block variable in nov152018_forkactivation.py

ptschip commented:
I removed the shadowing of "block"...hopefully that addresses your concerns.

On 04/10/2018 11:10 AM, awemany wrote:
>
> *@awemany* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In qa/rpc-tests/nov152018_forkactivation.py
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#discussion_r222774194>:
>
> > +        def update_block(block_number, new_transactions=[]):
> +            block = self.blocks[block_number]
> +            self.add_transactions_to_block(block, new_transactions)
> +            old_sha256 = block.sha256
> +            block.hashMerkleRoot = block.calc_merkle_root()
> +            block.solve()
> +            # Update the internal state just like in next_block
> +            self.tip = block
> +            if block.sha256 != old_sha256:
> +                self.block_heights[block.sha256] = self.block_heights[old_sha256]
> +                del self.block_heights[old_sha256]
> +            self.blocks[block_number] = block
> +            return block
> +
> +        # shorthand for functions
> +        block = self.next_block
>
> I find this pretty confusing. |block| is used both as a |CBlock| as
> well as the method |next_block()|. Wouldn't it be better to make this
> explicit?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#pullrequestreview-161742919>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0FSoeHUR88gVkEQSH9MpjPRriMlJks5uhk8LgaJpZM4XE1m2>.
>

ptschip commented:
unless i misunderstood you it already does...the very last test in the
script

# But large transactions are still ok.

On 04/10/2018 11:11 AM, awemany wrote:
>
> *@awemany* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In qa/rpc-tests/nov152018_forkactivation.py
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#discussion_r222774459>:
>
> > +        assert_equal(len(small_tx_block.vtx[1].serialize()), MIN_TX_SIZE - 1)
> +        yield accepted()
> +
> +        # Now MTP is exactly the fork time. Small transaction are now rejected.
> +        assert_equal(node.getblockheader(node.getbestblockhash())['mediantime'],
> +                     NOV152018_START_TIME)
> +
> +        # Now that the fork activated, it is not possible to have
> +        # small transactions anymore.
> +        bestblockhash = self.nodes[0].getbestblockhash()
> +        small_tx_block = block(4445, out[1], MIN_TX_SIZE - 1)
> +        assert_equal(len(small_tx_block.vtx[1].serialize()), MIN_TX_SIZE - 1)
> +        #check that this block was rejected.
> +        yield rejected()
> +        #yield rejected(RejectResult(16, b'bad-txns-undersize'))
> +        assert_equal(bestblockhash, self.nodes[0].getbestblockhash())
>
> Shouldn't this test the boundary - as in also make sure that
> |MIN_TX_SIZE| just barely goes through?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#pullrequestreview-161743258>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0LjzmC5xMBCgGCNthJJcG7OHZsGdks5uhk9CgaJpZM4XE1m2>.
>

awemany commented:
@ptschip: On the boundary test: Yes indeed, I missed that. I wonder whether it should be thematically sorted to be adjacent to the MIN_TX_SIZE-1 case.
On the shadowing: Didn't even see that it was shadowing, was rather confused by the naming scheme alone.

ptschip committed:
Add handling for IsNov152018Sheduled()

ptschip committed:
Fix activation test by using conensus.forkNov152018

gandrewstone commented on src/txadmission.cpp:
```diff
@@ -567,6 +567,13 @@ bool ParallelAcceptToMemoryPool(Snapshot &ss,
+    if (IsNov152018Scheduled() && IsNov152018Enabled(chainparams.GetConsensus(), chainActive.Tip()))
+    {
+        if (tx->GetTxSize() < MIN_TX_SIZE)
+            return state.DoS(0, false, REJECT_NONSTANDARD, "txn-undersize");
```
should be REJECT_INVALID

gandrewstone commented on src/validation/validation.cpp:
```diff
@@ -1406,6 +1406,14 @@ bool ContextualCheckBlock(const CBlock &block, CValidationState &state, CBlockIn
+        {
+            if (tx->GetTxSize() < MIN_TX_SIZE)
+                return state.DoS(10, error("%s: contains transactions that are too small", __func__), REJECT_INVALID,
+                    "bad-txns-undersize");
```
let's use the same error as was used above: "txn-undersize"

gandrewstone commented on src/validation/validation.cpp:
```diff
@@ -1743,6 +1751,14 @@ uint32_t GetBlockScriptFlags(const CBlockIndex *pindex, const Consensus::Params
 
+    // The Nov 15, 2018 HF enable sig push only and atrt enforcing also
+    // clean stack rules (see  BIP 62 for more details)
+    if (IsNov152018Enabled(consensusparams, pindex->pprev))
```
This is missing the IsNov152018Scheduled() call, which makes it inconsistent with other uses in this PR.  Its actually correct code because if ...Scheduled() returns false ...Enabled() will always return false.  But I think that for clarity we should add it here to make it consistent with the other uses.

But FYI you only really need to use both Scheduled and Enabled if you are checking whether you are before a scheduled fork.  For example, during a rollback you might need to check if the fork was on but is now off.  Another use is to override a configuration "tweak" (if you had one) to the pre or post-fork state based on the fork time.  You only want to override if the fork is actually scheduled.

ptschip commented:
done

On 09/10/2018 7:44 AM, Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/txadmission.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#discussion_r223732428>:
>
> > @@ -567,6 +567,13 @@ bool ParallelAcceptToMemoryPool(Snapshot &ss,
>      if (!CheckFinalTx(*tx, STANDARD_LOCKTIME_VERIFY_FLAGS))
>          return state.DoS(0, false, REJECT_NONSTANDARD, "non-final");
>  
> +    // Make sure tx size is acceptable after Nov 15, 2018 fork
> +    if (IsNov152018Scheduled() && IsNov152018Enabled(chainparams.GetConsensus(), chainActive.Tip()))
> +    {
> +        if (tx->GetTxSize() < MIN_TX_SIZE)
> +            return state.DoS(0, false, REJECT_NONSTANDARD, "txn-undersize");
>
> should be REJECT_INVALID
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#pullrequestreview-162930353>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0OS1BkOYlIH2h8OtICDbsuClm-9Sks5ujLZdgaJpZM4XE1m2>.
>

ptschip commented:
done

On 09/10/2018 7:46 AM, Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/validation/validation.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#discussion_r223733156>:
>
> > @@ -1406,6 +1406,14 @@ bool ContextualCheckBlock(const CBlock &block, CValidationState &state, CBlockIn
>              return state.DoS(
>                  10, error("%s: contains a non-final transaction", __func__), REJECT_INVALID, "bad-txns-nonfinal");
>          }
> +
> +        // Make sure tx size is acceptable after Nov 15, 2018 fork
> +        if (IsNov152018Scheduled() && IsNov152018Enabled(consensusParams, chainActive.Tip()))
> +        {
> +            if (tx->GetTxSize() < MIN_TX_SIZE)
> +                return state.DoS(10, error("%s: contains transactions that are too small", __func__), REJECT_INVALID,
> +                    "bad-txns-undersize");
>
> let's use the same error as was used above: "txn-undersize"
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#pullrequestreview-162931345>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0P8BnKrifzkzntun5gETMTKFtf5mks5ujLbBgaJpZM4XE1m2>.
>

ptschip commented:
done

On 09/10/2018 7:59 AM, Andrew Stone wrote:
>
> *@gandrewstone* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/validation/validation.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#discussion_r223738848>:
>
> > @@ -1743,6 +1751,14 @@ uint32_t GetBlockScriptFlags(const CBlockIndex *pindex, const Consensus::Params
>          flags |= SCRIPT_VERIFY_NULLFAIL;
>      }
>  
> +    // The Nov 15, 2018 HF enable sig push only and atrt enforcing also
> +    // clean stack rules (see  BIP 62 for more details)
> +    if (IsNov152018Enabled(consensusparams, pindex->pprev))
>
> This is missing the IsNov152018Scheduled() call, which makes it
> inconsistent with other uses in this PR. Its actually correct code
> because if ...Scheduled() returns false ...Enabled() will always
> return false. But I think that for clarity we should add it here to
> make it consistent with the other uses.
>
> But FYI you only really need to use both Scheduled and Enabled if you
> are checking whether you are before a scheduled fork. For example,
> during a rollback you might need to check if the fork was on but is
> now off. Another use is to override a configuration "tweak" (if you
> had one) to the pre or post-fork state based on the fork time. You
> only want to override if the fork is actually scheduled.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1376#pullrequestreview-162938579>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0KobVXhk9sm36bILFjy7zwJnbrWpks5ujLnFgaJpZM4XE1m2>.
>

ptschip committed:
Change a few reject codes and add check for IsNov152018Scheduled()

ptschip committed:
fix format
